[BUG]

When using Rapids plugin with Databricks pass through authentication doesn’t work.

when selecting from delta tables Security Token Expired” error is displayed. 

**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
Avoid posting private or sensitive data.
When creating Databricks cluster enable pass through authentication on standard cluster. Issue select statement on any delta table which is not on dbfs.

**Expected behavior**
A clear and concise description of what you expected to happen.
Pass through authentication should work.

**Environment details (please complete the following information)**
 - Environment : Databricks 8.2 runtime 
 
 - Spark configuration settings related to the issue
 Pass through authentication enabled on Databricks runtime environment 8.2

**Additional context**
Add any other context about the problem here.
All functionality works as intended when data is on dbfs

@jlowe You are right. 
I just used latest 21.12 snapshot jars and verified that https://github.com/NVIDIA/spark-rapids/issues/3903 is fixed.
However when writing dataframe to blob storage using passthrough auth, the issue still exists.

Here is my reproduce statement:
```
dfcustomer.write.mode("overwrite").format("delta").save("abfss://containerxxx@storagexxx.dfs.core.windows.net/delta/tablename")
```

It will fail with:
```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 25.0 failed 4 times, most recent failure: Lost task 1.3 in stage 25.0 (TID 488) (10.13.1.6 executor 0): java.util.concurrent.ExecutionException: com.databricks.backend.daemon.data.client.adl.AzureCredentialNotFoundException: Could not find ADLS Gen2 Token
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readPartFiles$7(GpuMultiFileReader.scala:820)
	at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readPartFiles$7$adapted(GpuMultiFileReader.scala:819)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readPartFiles$4(GpuMultiFileReader.scala:819)
	at com.nvidia.spark.rapids.Arm.closeOnExcept(Arm.scala:87)
	at com.nvidia.spark.rapids.Arm.closeOnExcept$(Arm.scala:85)
	at com.nvidia.spark.rapids.FilePartitionReaderBase.closeOnExcept(GpuMultiFileReader.scala:236)
	at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readPartFiles$1(GpuMultiFileReader.scala:804)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.FilePartitionReaderBase.withResource(GpuMultiFileReader.scala:236)
```

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala
@@ -22,7 +22,6 @@ import java.util.concurrent.{Callable, ConcurrentLinkedQueue, Future, LinkedBloc
 
 import scala.annotation.tailrec
 import scala.collection.JavaConverters._
-import scala.collection.immutable.HashSet
 import scala.collection.mutable.{ArrayBuffer, LinkedHashMap, Queue}
 import scala.math.max
 
@@ -318,16 +317,17 @@ abstract class MultiFileCloudPartitionReaderBase(
   private def initAndStartReaders(): Unit = {
     // limit the number we submit at once according to the config if set
     val limit = math.min(maxNumFileProcessed, files.length)
+    val tc = TaskContext.get
     for (i <- 0 until limit) {
       val file = files(i)
       // Add these in the order as we got them so that we can make sure
       // we process them in the same order as CPU would.
-      tasks.add(getThreadPool(numThreads).submit(getBatchRunner(file, conf, filters)))
+      tasks.add(getThreadPool(numThreads).submit(getBatchRunner(tc, file, conf, filters)))
     }
     // queue up any left to add once others finish
     for (i <- limit until files.length) {
       val file = files(i)
-      tasksToRun.enqueue(getBatchRunner(file, conf, filters))
+      tasksToRun.enqueue(getBatchRunner(tc, file, conf, filters))
     }
     isInitted = true
     filesToRead = files.length
@@ -337,15 +337,17 @@ abstract class MultiFileCloudPartitionReaderBase(
    * The sub-class must implement the real file reading logic in a Callable
    * which will be running in a thread pool
    *
+   * @param tc   task context to use
    * @param file file to be read
    * @param conf the Configuration parameters
    * @param filters push down filters
    * @return Callable[HostMemoryBuffersWithMetaDataBase]
    */
   def getBatchRunner(
-    file: PartitionedFile,
-    conf: Configuration,
-    filters: Array[Filter]): Callable[HostMemoryBuffersWithMetaDataBase]
+      tc: TaskContext,
+      file: PartitionedFile,
+      conf: Configuration,
+      filters: Array[Filter]): Callable[HostMemoryBuffersWithMetaDataBase]
 
   /**
    * Get ThreadPoolExecutor to run the Callable.
@@ -630,6 +632,7 @@ abstract class MultiFileCoalescingPartitionReaderBase(
    * The sub-class must implement the real file reading logic in a Callable
    * which will be running in a thread pool
    *
+   * @param tc     task context to use
    * @param file   file to be read
    * @param outhmb the sliced HostMemoryBuffer to hold the blocks, and the implementation
    *               is in charge of closing it in sub-class
@@ -641,10 +644,11 @@ abstract class MultiFileCoalescingPartitionReaderBase(
    *         result._2 is the bytes read
    */
   def getBatchRunner(
-    file: Path,
-    outhmb: HostMemoryBuffer,
-    blocks: ArrayBuffer[DataBlockBase],
-    offset: Long): Callable[(Seq[DataBlockBase], Long)]
+      tc: TaskContext,
+      file: Path,
+      outhmb: HostMemoryBuffer,
+      blocks: ArrayBuffer[DataBlockBase],
+      offset: Long): Callable[(Seq[DataBlockBase], Long)]
 
   /**
    * File format short name used for logging and other things to uniquely identity
@@ -807,13 +811,14 @@ abstract class MultiFileCoalescingPartitionReaderBase(
           var offset = writeFileHeader(hmb)
 
           val allOutputBlocks = scala.collection.mutable.ArrayBuffer[DataBlockBase]()
+          val tc = TaskContext.get
           filesAndBlocks.foreach { case (file, blocks) =>
             val fileBlockSize = blocks.map(_.getBlockSize).sum
             // use a single buffer and slice it up for different files if we need
             val outLocal = hmb.slice(offset, fileBlockSize)
             // Third, copy the blocks for each file in parallel using background threads
             tasks.add(getThreadPool(numThreads).submit(
-              getBatchRunner(file, outLocal, blocks, offset)))
+              getBatchRunner(tc, file, outLocal, blocks, offset)))
             offset += fileBlockSize
           }
 

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScanBase.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScanBase.scala
@@ -58,6 +58,7 @@ import org.apache.spark.sql.execution.datasources.v2.{EmptyPartitionReader, File
 import org.apache.spark.sql.execution.datasources.v2.orc.OrcScan
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.rapids.OrcFilters
+import org.apache.spark.sql.rapids.execution.TrampolineUtil
 import org.apache.spark.sql.sources.Filter
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.vectorized.ColumnarBatch
@@ -1296,6 +1297,7 @@ class MultiFileCloudOrcPartitionReader(
     requestedMapping: Option[Array[Int]]) extends HostMemoryBuffersWithMetaDataBase
 
   private class ReadBatchRunner(
+      taskContext: TaskContext,
       partFile: PartitionedFile,
       conf: Configuration,
       filters: Array[Filter]) extends Callable[HostMemoryBuffersWithMetaDataBase]  {
@@ -1303,6 +1305,15 @@ class MultiFileCloudOrcPartitionReader(
     private var blockChunkIter: BufferedIterator[OrcOutputStripe] = null
 
     override def call(): HostMemoryBuffersWithMetaDataBase = {
+      TrampolineUtil.setTaskContext(taskContext)
+      try {
+        doRead()
+      } finally {
+        TrampolineUtil.unsetTaskContext()
+      }
+    }
+
+    private def doRead(): HostMemoryBuffersWithMetaDataBase = {
       val startingBytesRead = fileSystemBytesRead()
 
       val hostBuffers = new ArrayBuffer[(HostMemoryBuffer, Long)]
@@ -1358,14 +1369,18 @@ class MultiFileCloudOrcPartitionReader(
    * The sub-class must implement the real file reading logic in a Callable
    * which will be running in a thread pool
    *
+   * @param tc      task context to use
    * @param file    file to be read
    * @param conf    the Configuration parameters
    * @param filters push down filters
    * @return Callable[HostMemoryBuffersWithMetaDataBase]
    */
-  override def getBatchRunner(file: PartitionedFile, conf: Configuration, filters: Array[Filter]):
-      Callable[HostMemoryBuffersWithMetaDataBase] = {
-    new ReadBatchRunner(file, conf, filters)
+  override def getBatchRunner(
+      tc: TaskContext,
+      file: PartitionedFile,
+      conf: Configuration,
+      filters: Array[Filter]): Callable[HostMemoryBuffersWithMetaDataBase] = {
+    new ReadBatchRunner(tc, file, conf, filters)
   }
 
   /**
@@ -1627,6 +1642,7 @@ class MultiFileOrcPartitionReader(
   // The runner to copy stripes to the offset of HostMemoryBuffer and update
   // the StripeInformation to construct the file Footer
   class OrcCopyStripesRunner(
+      taskContext: TaskContext,
       file: Path,
       outhmb: HostMemoryBuffer,
       stripes: ArrayBuffer[DataBlockBase],
@@ -1634,6 +1650,15 @@ class MultiFileOrcPartitionReader(
     extends Callable[(Seq[DataBlockBase], Long)] {
 
     override def call(): (Seq[DataBlockBase], Long) = {
+      TrampolineUtil.setTaskContext(taskContext)
+      try {
+        doRead()
+      } finally {
+        TrampolineUtil.unsetTaskContext()
+      }
+    }
+
+    private def doRead(): (Seq[DataBlockBase], Long) = {
       val startBytesRead = fileSystemBytesRead()
       // copy stripes to the HostMemoryBuffer
       withResource(outhmb) { _ =>
@@ -1800,6 +1825,7 @@ class MultiFileOrcPartitionReader(
    * The sub-class must implement the real file reading logic in a Callable
    * which will be running in a thread pool
    *
+   * @param tc     task context to use
    * @param file   file to be read
    * @param outhmb the sliced HostMemoryBuffer to hold the blocks, and the implementation
    *               is in charge of closing it in sub-class
@@ -1811,11 +1837,12 @@ class MultiFileOrcPartitionReader(
    *         result._2 is the bytes read
    */
   override def getBatchRunner(
+      tc: TaskContext,
       file: Path,
       outhmb: HostMemoryBuffer,
       blocks: ArrayBuffer[DataBlockBase],
       offset: Long): Callable[(Seq[DataBlockBase], Long)] = {
-    new OrcCopyStripesRunner(file, outhmb, blocks, offset)
+    new OrcCopyStripesRunner(tc, file, outhmb, blocks, offset)
   }
 
   /**

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScanBase.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScanBase.scala
@@ -1012,6 +1012,7 @@ class MultiFileParquetPartitionReader(
 
   // The runner to copy blocks to offset of HostMemoryBuffer
   class ParquetCopyBlocksRunner(
+      taskContext: TaskContext,
       file: Path,
       outhmb: HostMemoryBuffer,
       blocks: ArrayBuffer[DataBlockBase],
@@ -1019,16 +1020,21 @@ class MultiFileParquetPartitionReader(
     extends Callable[(Seq[DataBlockBase], Long)] {
 
     override def call(): (Seq[DataBlockBase], Long) = {
-      val startBytesRead = fileSystemBytesRead()
-      val res = withResource(outhmb) { _ =>
-        withResource(new HostMemoryOutputStream(outhmb)) { out =>
-          withResource(file.getFileSystem(conf).open(file)) { in =>
-            copyBlocksData(in, out, blocks, offset)
+      TrampolineUtil.setTaskContext(taskContext)
+      try {
+        val startBytesRead = fileSystemBytesRead()
+        val res = withResource(outhmb) { _ =>
+          withResource(new HostMemoryOutputStream(outhmb)) { out =>
+            withResource(file.getFileSystem(conf).open(file)) { in =>
+              copyBlocksData(in, out, blocks, offset)
+            }
           }
         }
+        val bytesRead = fileSystemBytesRead() - startBytesRead
+        (res, bytesRead)
+      } finally {
+        TrampolineUtil.unsetTaskContext()
       }
-      val bytesRead = fileSystemBytesRead() - startBytesRead
-      (res, bytesRead)
     }
   }
 
@@ -1074,11 +1080,12 @@ class MultiFileParquetPartitionReader(
   }
 
   override def getBatchRunner(
+      taskContext: TaskContext,
       file: Path,
       outhmb: HostMemoryBuffer,
       blocks: ArrayBuffer[DataBlockBase],
       offset: Long): Callable[(Seq[DataBlockBase], Long)] = {
-    new ParquetCopyBlocksRunner(file, outhmb, blocks, offset)
+    new ParquetCopyBlocksRunner(taskContext, file, outhmb, blocks, offset)
   }
 
   override final def getFileFormatShortName: String = "Parquet"
@@ -1197,7 +1204,9 @@ class MultiFileCloudParquetPartitionReader(
       hasInt96Timestamps: Boolean,
       clippedSchema: MessageType) extends HostMemoryBuffersWithMetaDataBase
 
-  private class ReadBatchRunner(filterHandler: GpuParquetFileFilterHandler,
+  private class ReadBatchRunner(
+      taskContext: TaskContext,
+      filterHandler: GpuParquetFileFilterHandler,
       file: PartitionedFile,
       conf: Configuration,
       filters: Array[Filter]) extends Callable[HostMemoryBuffersWithMetaDataBase] with Logging {
@@ -1213,6 +1222,15 @@ class MultiFileCloudParquetPartitionReader(
      * Note that the TaskContext is not set in these threads and should not be used.
      */
     override def call(): HostMemoryBuffersWithMetaData = {
+      TrampolineUtil.setTaskContext(taskContext)
+      try {
+        doRead()
+      } finally {
+        TrampolineUtil.unsetTaskContext()
+      }
+    }
+
+    private def doRead(): HostMemoryBuffersWithMetaData = {
       val startingBytesRead = fileSystemBytesRead()
       val hostBuffers = new ArrayBuffer[(HostMemoryBuffer, Long)]
       try {
@@ -1271,14 +1289,18 @@ class MultiFileCloudParquetPartitionReader(
   /**
    * File reading logic in a Callable which will be running in a thread pool
    *
+   * @param tc      task context to use
    * @param file    file to be read
    * @param conf    configuration
    * @param filters push down filters
    * @return Callable[HostMemoryBuffersWithMetaDataBase]
    */
-  override def getBatchRunner(file: PartitionedFile, conf: Configuration, filters: Array[Filter]):
-      Callable[HostMemoryBuffersWithMetaDataBase] = {
-    new ReadBatchRunner(filterHandler, file, conf, filters)
+  override def getBatchRunner(
+      tc: TaskContext,
+      file: PartitionedFile,
+      conf: Configuration,
+      filters: Array[Filter]): Callable[HostMemoryBuffersWithMetaDataBase] = {
+    new ReadBatchRunner(tc, filterHandler, file, conf, filters)
   }
 
   /**

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala
@@ -145,4 +145,10 @@ object TrampolineUtil {
 
   /** Throw a Spark analysis exception */
   def throwAnalysisException(msg: String) = throw new AnalysisException(msg)
+
+  /** Set the task context for the current thread */
+  def setTaskContext(tc: TaskContext): Unit = TaskContext.setTaskContext(tc)
+
+  /** Remove the task context for the current thread */
+  def unsetTaskContext(): Unit = TaskContext.unset()
 }

--- a/tests/src/test/scala/com/nvidia/spark/rapids/GpuMultiFileReaderSuite.scala
+++ b/tests/src/test/scala/com/nvidia/spark/rapids/GpuMultiFileReaderSuite.scala
@@ -22,6 +22,7 @@ import ai.rapids.cudf.HostMemoryBuffer
 import org.apache.hadoop.conf.Configuration
 import org.scalatest.FunSuite
 
+import org.apache.spark.TaskContext
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.execution.datasources.PartitionedFile
 import org.apache.spark.sql.sources.Filter
@@ -49,6 +50,7 @@ class GpuMultiFileReaderSuite extends FunSuite with Arm {
       })
 
       override def getBatchRunner(
+          tc: TaskContext,
           file: PartitionedFile,
           conf: Configuration,
           filters: Array[Filter]): Callable[HostMemoryBuffersWithMetaDataBase] = {

